shcomplete/tfdf.py

```diff
@@ -0,0 +1,112 @@
+import logging
+import os
+import re
+from math import log
+from clint.textui import progress
+
+import pygtrie
+import numpy
+
+
+def get_tries(args, _log):
+    """
+    Return a list of tries, one for each file where the keys are the command lines.
+    """
+    tries = []
+    max_len = args.max_length
+    for root, _, files in os.walk(args.directory):
+        for file_name in files:
+            path_to_file = os.path.join(root, file_name)
+            _log.info("parsing %s", path_to_file)
+            trie = pygtrie.StringTrie(separator=" ")
+            tries.append((path_to_file, trie))
+            with open(path_to_file, "r") as f:
+                lines = f.readlines()
+                for line in lines:
+                    line = [l for l in line.rstrip().replace("\t", " ").split(" ") if l][:max_len]
+                    key = " ".join(line)
+                    if key not in trie:
+                        trie[key] = 0
+                    prefixes = []
+                    for i in range(1, len(line)+1):
+                        prefixes.append(" ".join(line[:i]))
+                    for pfx in prefixes:
+                        trie[pfx] = trie.get(pfx, 0) + 1
+    return tries
+
+
+def get_df_trie(tries):
+    """
+    Return a trie where the prefix's value is its document frequency among the tries.
+    """
+    df_trie = pygtrie.StringTrie(separator=" ")
+    for _, trie in progress.bar(tries, expected_size=len(tries)):
+        for prefix, _ in trie.iteritems():
+            if prefix not in df_trie:
+                df_trie[prefix] = 0
+            df_trie[prefix] += 1
+    return df_trie
+
+
+def sum_tfdf_tries(tries, df_trie, _log):
+    """
+    Compute the TF-DF scores of each prefix in each trie.
+    Sum these scores in one big trie that is returned.
+    """
+    big_trie = pygtrie.StringTrie(separator=" ")
+    for path_to_file, trie in tries:
+        _log.info("%s %d", path_to_file, len(trie))
+        stack = [("", trie._root, df_trie._root)]
+        while stack:
+            path, local_node, df_node = stack.pop(0)
+            for child_key, child_node in local_node.children.items():
+                stack.append((path + " " + child_key, child_node, df_node.children[child_key]))
+                try:
+                    tfdf = (local_node.value / len(trie)) * ((df_node.value - 1) / len(tries))
+                    big_trie[path[1:]] = big_trie.get(path[1:], 0) + tfdf
```

Comment:
Better:

```
stack.append((((path + " ") if path else "") + child_key, ...))
# ...
big_trie[path] = big_trie.get(path, 0) + tfdf
```

This substring operation is rather expensive since it creates a string copy.